src/libcephfs_proxy/proxy_link.c

```diff
@@ -55,6 +68,523 @@ static int32_t proxy_link_prepare(struct sockaddr_un *addr, const char *path)
+		neg->v0.size = NEG_VERSION_SIZE(PROXY_LINK_NEGOTIATE_VERSION);
+	}
+
+	if (negotiation_sizes[neg->v0.version] != neg->v0.size) {
+		return proxy_log(LOG_ERR, EINVAL,
+				 "Negotiation structure doesn't have the "
+				 "expected size");
+	}
+
+	if (neg->v0.version < neg->v0.min_version) {
+		return proxy_log(LOG_ERR, ENOTSUP,
+				 "Negotiation requires an unsupported version");
+	}
+
+	/* Read remaining negotiation data. */
+	err = proxy_link_read(link, sd, ptr, neg->v0.size - sizeof(neg->v0));
+	if (err < 0) {
+		return err;
+	}
+
+	/* Ignore any additional data belonging to unsupported negotiation
+	 * structures. */
+	size -= neg->v0.size;
+	while (size > 0) {
+		len = sizeof(buffer);
+		if (len > size) {
+			len = size;
+		}
+		err = proxy_link_read(link, sd, buffer, len);
+		if (err < 0) {
+			return err;
+		}
+		size -= len;
+	}
+
+	return 0;
+}
+
+static int32_t proxy_link_negotiate_check(proxy_link_negotiate_t *local,
+					  proxy_link_negotiate_t *remote,
+					  proxy_link_negotiate_cbk_t cbk)
+{
+	uint32_t supported, enabled;
+	int32_t err;
+
+	if (local->v0.version > remote->v0.version) {
+		local->v0.version = remote->v0.version;
+		local->v0.size = remote->v0.size;
+	}
+
+	if (remote->v0.version == 0) {
+		/* Legacy peer. If we require any feature, the peer won't
+		 * support it, so we can't continue */
+		if (local->v1.required != 0) {
+			return proxy_log(LOG_ERR, ENOTSUP,
+					 "The peer doesn't support any "
+					 "features");
+		}
+
+		/* The peer is running the old version, but it is compatible
+		 * with us, so everything is fine and the connection can be
+		 * completed successfully with all features disabled. */
+
+		local->v1.enabled = 0;
+
+		proxy_log(LOG_INFO, 0,
+			  "Connected to legacy peer. No features enabled");
+
+		goto validate;
+	}
+
+	supported = local->v1.supported & remote->v1.supported;
+	local->v1.supported = supported;
+
+	if ((local->v1.required & ~supported) != 0) {
+		return proxy_log(LOG_ERR, ENOTSUP,
+				 "Required features are not supported by the "
+				 "peer");
+	}
+	if ((remote->v1.required & ~supported) != 0) {
+		return proxy_log(LOG_ERR, ENOTSUP,
+				 "The peer requires some features that are "
+				 "not supported");
+	}
+
+	/* For now just combine the desired features from each side. In the
+	 * future, if there are two features to implement the same, they may
+	 * be chosen here and enable just one of them. */
+
+	enabled = (local->v1.enabled | remote->v1.enabled) & supported;
+	local->v1.enabled = enabled;
+
+	/* NEG_VERSION: Implement handling of negotiate extensions. */
+
+validate:
+	if (cbk != NULL) {
+		err = cbk(local);
+		if (err < 0) {
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+static int32_t proxy_link_negotiate_client(proxy_link_t *link, int32_t sd,
+					   proxy_link_negotiate_t *neg,
+					   proxy_link_negotiate_cbk_t cbk)
+{
+	proxy_link_negotiate_t remote;
+	int32_t err;
+
+	err = proxy_link_write(link, sd, neg, neg->v0.size);
+	if (err < 0) {
+		return err;
```

Comment:
If there's failure shouldn't we consider to `close(sd)` ? 
```suggestion
		close(sd);
		return err;
```

Reply:
Failures while processing the negotiation are propagated to the callers. In this particular case, the error will be propagated up to one of these functions: `ceph_create()` or `proxy_global_connect()` in `libcephfs_proxy.c`. Both will take care of closing the connection if the negotiation fails by calling proxy_link_close().

In general, I think it's better to release resources at the same level they have been acquired, which means that the function that actually opens the connection should be the responsible of closing it in case that something fails. The negotiation functions receive an already open connection, so they should return without closing it, even in case of error.